src/ccache.c

```diff
@@ -0,0 +1,523 @@
+	}
+	assert(ret != NULL);
+	if (unlikely(zero)) {
+		size_t usize = sz_index2size(ind);
+		assert(tcache_salloc(tsd_tsdn(tsd), ret) == usize);
+		memset(ret, 0, usize);
+	}
+	ccache_bump_tstats(tsd, ind);
+
+	return ret;
+}
+
+bool
+ccache_free(tsd_t *tsd, void *ptr, szind_t ind, bool small) {
+	edata_t *edata = emap_edata_lookup(tsd_tsdn(tsd),
+	    &arena_emap_global, ptr);
+	arena_t *origin_arena = arena_get_from_edata(edata);
+	if (!arena_is_auto(origin_arena)) {
+		return false;
+	}
+	if (!ccache_handles_szind(ind)) {
+		return false;
+	}
+	volatile rseq_t *rseq_abi = &tsd_ccache_tdatap_get(tsd)->rseq_abi;
+	int cpu;
+	ccache_t *ccache;
+	ccache_bin_t *bin;
+	void *full_mark;
+restart:
+	cpu = rseq_abi->cpu_id;
+	ccache = ccache_get(cpu);
+	bin = ccache_bin_get(ccache, ind);
+	full_mark = bin->ccache_bin_entry;
+	__asm__ __volatile__ goto(
+	    /* aw == Allocatable and writable section */
+	    ".pushsection __rseq_table, \"aw\"\n\t"
+	    ".balign 32\n\t"
+	    "cs_obj%=:\n\t"
+	    /* rseq_cs.version rseq_cs.flags */
+	    ".long 0, 0\n\t"
+	    /* rseq_cs.start_ip rseq_cs.post_commit_offset rseq_cs.abort_ip */
+	    ".quad rseq_cs_start%=, rseq_cs_postcommit%= - rseq_cs_start%=, rseq_cs_abort%=\n\t"
+	    ".popsection\n\t"
+
+	    "leaq cs_obj%=(%%rip), %%rax\n\t"
+	    "movq %%rax, %[rseq_cs]\n\t"
+
+	    /* -------------------- Begin ------------------- */
+	    "rseq_cs_start%=:\n\t"
+	    "cmpl %[cpu_id], %[initial_cpu_id]\n\t"
+	    /* Restart in case of CPU id missmatch*/
+	    "jne %l[restart]\n\t"
+
+	    /* rax = head */
+	    "movq %c[head_offset](%[bin]), %%rax\n\t"
+	    /* Check if in fallback mode, then fallback */
+	    "test %%rax, %%rax\n\t"
+	    "je %l[fallback]\n\t"
+	    /* Check if the bin is full, then flush  */
+	    "cmpq %%rax, %[full_mark]\n\t"
+	    "jne flush_not_needed%=\n\t"
+	    /* Prepare head = 0 */
+	    "xor %%rax, %%rax\n\t"
+	    "jmp commit%=\n\t"
+
+	    /* Find head-1, write ptr there. Commit head-1 by writing reg to bin */
+	    "flush_not_needed%=:\n\t"
+	    /* Prepare head-- */
+	    "subq $8, %%rax\n\t"
+	    "movq %[ptr], (%%rax)\n\t"
+	    /* Commit either head-- or head = 0 */
+	    "commit%=:\n\t"
+	    "movq %%rax, %c[head_offset](%[bin])\n\t"
+
+	    /* ------------------- End ------------------- */
+	    "rseq_cs_postcommit%=:\n\t"
+
+	    /* ax == Allocatable and executable section */
+	    ".pushsection __rseq_abort, \"ax\"\n\t"
+	    ".byte 0x0f, 0x1f, 0x05\n\t"
+	    ".long "STRINGIFY(JEMALLOC_RSEQ_SIG)"\n\t"
+	    "rseq_cs_abort%=:\n\t"
+	    "jmp rseq_cs_start%=\n\t"
+	    ".popsection\n\t"
+	    :
+	    : [bin] "r" (bin),
+	      [head_offset] "i" (offsetof(ccache_bin_t, head)),
+	      [full_mark] "r" (full_mark),
+	      [rseq_cs] "m" (rseq_abi->rseq_cs),
+	      [cpu_id] "m" (rseq_abi->cpu_id_start),
+	      [initial_cpu_id] "r" (cpu),
+	      [ccache] "rm" (ccache),
+	      [ptr] "r" (ptr)
+	    : "rax", "cc"
+	    : restart, fallback
+	 );
+
+	if (bin->head == NULL) {
```

Comment:
Question: it looks this indicates both refill and flush mode. If the flushing thread gets preempted here, and another thread does a lot of activities then trigger a refill -- would that be an issue? e.g. ccache_bin_flush may encounter an empty bin, or refill encounter a full bin.

Reply:
It should be impossible.
1) There is only one thread attempting to alloc/free from a specific cache bin at a time, in a critical section.
2) The result of the critical section is either preemption (no side effects) or a commit, which atomically changes stack->head to one of the following states: (+=1; -=1; =0).
3) Only the thread, which has changed head to 0, is capable of reaching `if (bin->head == NULL)`. Then it's the owner of the refilling/flushing.
4) Other threads, scheduled on the same CPU and entering the critical section, will see `head == null` by executing `test %%rax, %%rax` and realize that the bin is in the fallback mode, therefore there must be another thread that has put the bin into this state, and that thread is responsible for refilling/flushing the bin. Thus, the current thread must not try allocating from the ccache.
5) Once the thread, responsible for refilling/flushing, finishes the procedure, it makes the bin accessible to other threads by storing the non-zero value to `bin->head`.

Reply:
Actually, you were right, this was a bug. Sorry for the misleading previous reply.

Good thing is that it was caught by the fuzzy test. Making the scheduler switch more aggressively also helps.